### PR TITLE
Add unload method to MLModel interface

### LINKED ISSUE #1145: Expose `unload` api  in `MLModel`
Currently there is no explicit user-defined unload utility that the user can define. This is required in cases where the user wants to do explicit clean up when the mlserver model is unloaded. e.g. closing connections / releasing resources etc.

### PR DESCRIPTION
**Pull request by adriangonz**
Fixes #1145 

## Changelog

- Add `unload()` method in `MLModel` interface (called by registry when unloading models)

## TODO

- [x] In HF runtime, unload model from GPU (as discussed in https://github.com/SeldonIO/MLServer/issues/986#issuecomment-1587422911)

### CODE REVIEW

**Review by sakoush — APPROVED** [review]
LGTM. I left some comments for potential improvements.

File: docs/user-guide/custom.md [hunk 1]
```diff
@@ -20,6 +20,9 @@ abstract class, whose main methods are:
 - {func}`load() <mlserver.MLModel.load>`:
   Responsible for loading any artifacts related to a model (e.g. model
   weights, pickle files, etc.).
+- {func}`load() <mlserver.MLModel.unload>`:
```
[the remark this answers is not in the document]
Comment (adriangonz) [reply docs/user-guide/custom.md]: Good spot!

File: mlserver/registry.py [hunk 2]
```diff
@@ -186,6 +186,7 @@ async def _reload_model(self, old_model: MLModel, new_model: MLModel):
         if old_model == self.default:
             self._clear_default()
 
+        old_model.ready = not await old_model.unload()
```
[the remark this answers is not in the document]
Comment (adriangonz) [reply mlserver/registry.py:189]: This is just semantics. The `unload()` method returns a boolean which is `True` when everything went OK. Therefore, if the `unload()` method had no troubles, then the model is no longer ready (i.e. `ready = not unload()`). 

Happy to remove the boolean return value though - I added it mainly for parity with the `load()` method, but maybe it doesn't make that much sense for `unload`?

File: runtimes/huggingface/mlserver_huggingface/runtime.py [hunk 3]
```diff
@@ -50,6 +51,20 @@ async def predict(self, payload: InferenceRequest) -> InferenceResponse:
...
+        if not is_torch:
+            return True
+
+        uses_gpu = torch.cuda.is_available() and self._model.device != -1
+        if not uses_gpu:
+            # Nothing to free
+            return True
+
+        # Free up Torch's GPU memory
+        torch.cuda.empty_cache()
```
[the remark this answers is not in the document]
Comment (adriangonz) [reply runtimes/huggingface/mlserver_huggingface/runtime.py:65]: Good point - let me add that

**Pull request merged**